fix #18810 Staff type change issues

sammik opened the pull request:
Resolves: #18810  <!-- Replace `NNNNN` with a GitHub issue number, or a direct link if the issue is not on GitHub -->

<!-- Add a short description of and motivation for the changes here -->

https://github.com/musescore/MuseScore/assets/1646034/23364246-1325-4d60-b208-5801f5c7ee51



<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

sammik commented:
@oktophonie

sammik commented:
@cbjeukendrup please, could you look into, why utests failes?

I cannot find useful info in log.

Thank you!

Jojo-Schmitz commented:
Then look again:
```
1: [ RUN      ] Global_IO_IODeviceTests.Open_ReadOnly
1: 09:01:33.542 | INFO  | main_thread | GlobalModule | onPreInit: log path: /root/.local/share/global_tests/logs/MuseScore_230822_090133.log
1: 09:01:33.542 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.2.0-dev, build number 232340824 ===
1: 09:01:33.544 | ERROR | main_thread | IODevice   | write: "ASSERT FAILED!": isOpenModeWriteable(), file: /home/runner/work/MuseScore/MuseScore/src/framework/global/io/iodevice.cpp, line: 200
1: [       OK ] Global_IO_IODeviceTests.Open_ReadOnly (408 ms)
1: [ RUN      ] Global_IO_IODeviceTests.Open_WriteOnly
1: 09:01:33.949 | INFO  | main_thread | GlobalModule | onPreInit: log path: /root/.local/share/global_tests/logs/MuseScore_230822_090133.log
1: 09:01:33.949 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.2.0-dev, build number 232340824 ===
1: 09:01:33.951 | ERROR | main_thread | IODevice   | read: "ASSERT FAILED!": isOpenModeReadable(), file: /home/runner/work/MuseScore/MuseScore/src/framework/global/io/iodevice.cpp, line: 152
1: 09:01:34.216 | INFO  | main_thread | GlobalModule | onPreInit: log path: /root/.local/share/global_tests/logs/MuseScore_230822_090134.log
1: 09:01:34.216 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.2.0-dev, build number 232340824 ===
1: 09:01:34.217 | ERROR | main_thread | IODevice   | read: "ASSERT FAILED!": isOpenModeReadable(), file: /home/runner/work/MuseScore/MuseScore/src/framework/global/io/iodevice.cpp, line: 152
1: 09:01:34.480 | INFO  | main_thread | GlobalModule | onPreInit: log path: /root/.local/share/global_tests/logs/MuseScore_[230](https://github.com/musescore/MuseScore/actions/runs/5936302532/job/16096440598?pr=19086#step:9:231)822_090134.log
1: 09:01:34.480 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.2.0-dev, build number 232340824 ===
1: 09:01:34.482 | ERROR | main_thread | IODevice   | read: "ASSERT FAILED!": isOpenModeReadable(), file: /home/runner/work/MuseScore/MuseScore/src/framework/global/io/iodevice.cpp, line: 126
```

cbjeukendrup commented:
@Jojo-Schmitz Those messages are unrelated and expected; that test is a special one that checks whether the ASSERT happens correctly. So that test actually passes if MuseScore successfully crashes.

@sammik The reason for the failure is that the test run exceeded the maximum duration of 360 minutes. That might mean that there was a temporary bug in GitHub Actions, so I will try if restarting the check helps; but it might also mean that there is an infinite loop in one of the test cases, so you could try on your own computer if that's the case.

Jojo-Schmitz commented:
😬 oops

sammik commented:
@cbjeukendrup thanks, i was thinking, if infinite loop could be case
Ill try run it locally.

oktophonie commented:
Regarding the issue that this PR resolves: points 2 and 3 seem to be fixed, as far as I can tell!

Point 1 (layout) is improved, except for this (somewhat hypothetical case):
![image](https://github.com/musescore/MuseScore/assets/20604032/08a0d50e-4fd2-42e4-893e-035b8f261e0b)
I feel like the barline in red should be the same as the one in green. That is: at any staff type change, the height of the barline should be the combined range of the barlines required by the staff type on either side. At the moment that's only a problem (in that it's inconsistent) when coming from a 1-line stave to any other type, but that may not always be the case - there have been requests for options to have taller barlines on a 2-line stave, for example.

(One day, this may all be rendered moot when we can have a 1-line stave aligned with the middle line of a 5-line stave, rather than the top line, but that's some way off; see https://github.com/musescore/MuseScore/issues/15697#issuecomment-1372434842.)

oktophonie commented:
One error still: note the initial system barline now!

![image](https://github.com/musescore/MuseScore/assets/20604032/327de83b-c8d3-4222-9a7f-e0eccc93b60e)

MarcSabatella commented:
Note you *can* have a one-line staff aligned in the middle, by assigning an offset of 2 sp in the staff type change:

![image](https://github.com/musescore/MuseScore/assets/1799936/e3f44b88-c043-4f83-b7a3-7b62044209d8)

This works as expected to me in terms of barline heights, both in current versions and with this PR.

cbjeukendrup commented on src/engraving/dom/edit.cpp:
```diff
@@ -2993,6 +2993,21 @@ void Score::deleteMeasures(MeasureBase* mbStart, MeasureBase* mbEnd, bool preser
+                            score->deleteItem(el);
+                        }
+                    }
+                    break;
```
Shouldn't this be two lines above instead? (Now, it won't check other staves after finding a staff type change on one staff; but perhaps you meant to stop checking other measure elements after finding the staff type change)

cbjeukendrup replied:
@sammik Sorry, I'm afraid I don't entirely understand what you mean; but does it look okay now?
(I pushed 1 commit containing fixes for these two review comments, and also did a rebase (but for the rest didn't touch your commits)

sammik commented:
@cbjeukendrup I rebased this PR, included Your "break one line up" review in "move staff type changes" and simplified "delete staffTypeChange" and moved it from "deleteMeasures" to "undoRemoveMeasures".

But figured out, there is one other issue related to split/join measures containing staffTypeChange, which is regression caused by this PR. Should I commit fix for it here, or better leave it, as it is, and create new issue / PR, after this will be merged?

Thank You for Your patience.

cbjeukendrup commented:
@sammik If it would be a regression, then it might be better to fix it immediately in this PR, but preferably as a separate commit. Is that okay for you?

sammik commented:
> in this PR, but preferably as a separate commit

@cbjeukendrup done

cbjeukendrup commented:
@sammik @oktophonie Do you think it needs to be re-tested, or can it be merged?

sammik commented:
If retested, only two things nas changed:
- removing staff type change in removed measures (so staff type should be same as it was in preceding measure)
- split / merge measure shouldn't affect staff type changes in following measures

Split / merge measure containing staff type change still removes staff type change, which is incorrect, but it is out of scope of this PR.
